src/DeprefixingContainer.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+namespace Dhii\Container;
+
+use Psr\Container\ContainerInterface;
+use Psr\Container\NotFoundExceptionInterface;
+
+/**
+ * A container implementation that wraps around an inner container to automatically add prefixes to keys during
+ * fetching and look up, allowing consumers to omit them.
+ *
+ * @since [*next-version*]
+ */
+class DeprefixingContainer implements ContainerInterface
+{
+    /**
+     * @since [*next-version*]
+     *
+     * @var ContainerInterface
+     */
+    protected $inner;
+
+    /**
+     * @since [*next-version*]
+     *
+     * @var string
+     */
+    protected $prefix;
+
+    /**
+     * @since [*next-version*]
+     *
+     * @var bool
+     */
+    protected $strict;
+
+    /**
+     * Constructor.
+     *
+     * @since [*next-version*]
+     *
+     * @param ContainerInterface $container The container whose keys to deprefix.
+     * @param string             $prefix    The prefix to remove from the container's keys.
+     * @param bool               $strict    Whether or not to fallback to prefixed keys if an un-prefixed key does not
+     *                                      exist in the inner container.
+     */
+    public function __construct(ContainerInterface $container, string $prefix, bool $strict = true)
+    {
+        $this->inner = $container;
+        $this->prefix = $prefix;
+        $this->strict = $strict;
+    }
+
+    /**
+     * @inheritdoc
+     *
+     * @since [*next-version*]
+     */
+    public function get($key)
+    {
+        try {
+            return $this->inner->get($this->getInnerKey($key));
+        } catch (NotFoundExceptionInterface $nfException) {
+            if ($this->strict) {
```

Comment:
This check is throwing off error reporting. If the inner container fails to get a service for the prefixed key, in non-strict mode it will attempt to fetch the non-prefixed key. If that also fails, the reported exception will be with regards to the latter, even though the consumer is most likely interested in the former.

Furthermore, if the container _does_ have a service for the prefixed key but fails to fetch one of its dependencies, this error reporting will be even more vague.

Consider service `foo` depends on `bar` depends on `baz`. If the container is asked to fetch `prefix/foo` and fails to find dependency `baz`, the exception will be caught by this `try..catch` block. In non-strict mode, the container will then attempt to fetch `foo` with no prefix and report that it couldn't find `foo`, when in reality it couldn't find `baz`.

This can be mitigated by a simple `has` check that does not alter the functionality of the container, but greatly improves error reporting.

```php
if (!$this->strict || !$this->inner->has($key)) {
	throw $nfException;
}
```